apps/src/lib/node/ledger/shell/governance.rs

```diff
@@ -0,0 +1,195 @@
+use anoma::ledger::governance::storage as gov_storage;
+use anoma::ledger::governance::utils::{
+    compute_tally, get_proposal_votes, ProposalEvent,
+};
+use anoma::ledger::governance::vp::ADDRESS as gov_address;
+use anoma::ledger::storage::types::encode;
+use anoma::ledger::storage::{DBIter, StorageHasher, DB};
+use anoma::ledger::treasury::ADDRESS as treasury_address;
+use anoma::types::address::{xan as m1t, Address};
+use anoma::types::governance::TallyResult;
+use anoma::types::storage::Epoch;
+use anoma::types::token;
+
+use super::*;
+use crate::node::ledger::events::EventType;
+
+pub struct ProposalsResult {
+    passed: Vec<u64>,
+    rejected: Vec<u64>,
+}
+
+pub fn execute_governance_proposals<D, H>(
+    shell: &mut Shell<D, H>,
+    new_epoch: bool,
+    response: &mut shim::response::FinalizeBlock,
+) -> Result<ProposalsResult>
+where
+    D: DB + for<'iter> DBIter<'iter> + Sync + 'static,
+    H: StorageHasher + Sync + 'static,
+{
+    let mut proposals_result = ProposalsResult {
+        passed: Vec::new(),
+        rejected: Vec::new(),
+    };
+
+    if !new_epoch {
+        return Ok(proposals_result);
+    }
+
+    for id in std::mem::take(&mut shell.proposal_data) {
+        let proposal_funds_key = gov_storage::get_funds_key(id);
+        let proposal_start_epoch_key =
+            gov_storage::get_voting_start_epoch_key(id);
+
+        let funds = shell
+            .read_storage_key::<token::Amount>(&proposal_funds_key)
+            .ok_or_else(|| {
+                Error::BadProposal(id, "Invalid proposal funds.".to_string())
+            })?;
+        let proposal_start_epoch = shell
+            .read_storage_key::<Epoch>(&proposal_start_epoch_key)
+            .ok_or_else(|| {
+                Error::BadProposal(
+                    id,
+                    "Invalid proposal start_epoch.".to_string(),
+                )
+            })?;
+
+        let votes =
+            get_proposal_votes(&shell.storage, proposal_start_epoch, id);
```

Comment:
should the epoch given to `get_proposal_votes` and `compute_tally` (here and other places where we're using these) be end epoch? (https://specs.anoma.net/master/architecture/governance/signaling.html#tally says it should be `end_epoch`)